chain/ethereum/src/data_source.rs

```diff
@@ -358,13 +404,40 @@ impl DataSource {
     fn handler_for_block(
         &self,
         trigger_type: &EthereumBlockTriggerType,
+        block: BlockNumber,
     ) -> Option<MappingBlockHandler> {
         match trigger_type {
-            EthereumBlockTriggerType::Every => self
+            // Start matches only initialization handlers with a `once` filter
+            EthereumBlockTriggerType::Start => self
                 .mapping
                 .block_handlers
                 .iter()
-                .find(move |handler| handler.filter.is_none())
+                .find(move |handler| match handler.filter {
+                    Some(BlockHandlerFilter::Once) => block == self.start_block,
+                    _ => false,
+                })
+                .cloned(),
+            // End matches all handlers without a filter or with a `polling` filter
+            EthereumBlockTriggerType::End => self
+                .mapping
+                .block_handlers
+                .iter()
+                .find(move |handler| match handler.filter {
+                    Some(BlockHandlerFilter::Polling { every }) => {
+                        // POLLING_TODO: Get creation block if start_block is zero
```

Comment:
This is very well noticed, and we need to fix before merging. There are two places where we set `start_block: 0` for dynamic data sources, maybe we should just set that to `start_block: creation_block` and this code would work fine.

Reply:
@leoyvens done!